File: 0696-count-binary-substrings/0696-count-binary-substrings.go
func countBinarySubstrings(s string) int {
    ans := 0
    for i := 0; i < len(s); i++ {
        count := 0
        curr := s[i]
        j := i
        for j < len(s) && s[j] == curr {
            count++
            j++
        }
        
        found := false
        for j < len(s) && s[j] != curr {
            count--
            j++
            if count == 0 {
                found = true
                break
            }
        }

        if found { ans++ }
    }

    return ans
}

// string, two pointers
// time: O(n)
// space: O(1)

File: 0696-count-binary-substrings/0696-count-binary-substrings.py
class Solution:
    def countBinarySubstrings(self, s: str) -> int:
        groups = [1]
        i = 0
        for i in range(1, len(s)):
            if s[i-1] != s[i]:
                groups.append(1)
                continue
            groups[-1] += 1
        
        ans = 0
        for i in range(1, len(groups)):
            ans += min(groups[i-1], groups[i])
        
        return ans


# string
# time: O(n)
# space: O(1)